Introduce NullMarked to some packages in spring-integration-core module

cppwfs opened the pull request:
This is a first pass at the module to make sure it is being converted properly.

There are 3 commits that comprise this PR.  Each commit handles a specific series of package(s) on the model.  
This was done to simplify the review process so each package could be reviewed indepently.
* Commit 1 - org.springframework.integration.container
* Commit 2 - org.springframework.integration.messaging.core
* Commit 3 - org.springframework.integration.aggregator & org.springframework.integration.acks

> [!NOTE]  
> Highlights information that users should take into account, even when skimming. in this commit that there cases where  methods are called where the parameters passed in may contain a null.
> In those cases the attribute that contains the potential null is tested with an `Assert.notNull`.  This may not be the best approach, let's discuss.

> [!NOTE]  
> Also note that in FluxAggregatorMessageHandler line 112 `@NullUnmarked` is used on the the applyWindowOptions method.
> This is because the consumer.apply on line 121 can not be null according to JSpecify.   But we can't make that assumption for how the user implemented the fuction.
>There maybe a better way of handling this besides marking the method as `NullUnmarked`.   Let's discuss.
> Similarly in FluxAggregatorMessageHandler the way in which sequenceSizeHeader is utilized it had to be `@NullUnmarked`.   Let's discuss

Related to https://github.com/spring-projects/spring-integration/issues/10083

cppwfs committed:
Update SI Core to add Nullability to module

This is a first pass at the module to make sure it is being converted properly.
Related to https://github.com/spring-projects/spring-integration/issues/10083

cppwfs committed:
Add nullability to org.springframework.messaging.core package

cppwfs committed:
Add nullability to org.springframework.aggregator & org.springframework.acks packages

Note in this commit that there cases where methods are called where the parameters passed in may contain a null.
In those cases the attribute that contains the potential null is tested with an `Assert.notNull`.  This may not be the best approach, let's discuss.

Also note that in FluxAggregatorMessageHandler line 112 `@NullUnmarked` is used on the the applyWindowOptions method.
This is because the consumer.apply on line 121 can not be null according to JSpecify.   But we can't make that assumption for how the user implemented the function.
There maybe a better way of handling this besides marking the method as `NullUnmarked`.   Let's discuss.

Similarly in FluxAggregatorMessageHandler the way in which sequenceSizeHeader is utilized it had to be `@NullUnmarked`.   Let's discuss

Make sure all annotations are applied consistently

artembilan commented on spring-integration-core/src/main/java/org/springframework/integration/aggregator/FluxAggregatorMessageHandler.java:
```diff
@@ -62,18 +64,24 @@ public class FluxAggregatorMessageHandler extends AbstractMessageProducingHandle
 
-	private Function<Message<?>, Integer> windowSizeFunction = FluxAggregatorMessageHandler::sequenceSizeHeader;
+	@Nullable
+	public Function<Message<?>, Integer> windowSizeFunction = FluxAggregatorMessageHandler::sequenceSizeHeader;
```
This is not correct. The `@Nullable` on the property type would mean that `FluxAggregatorMessageHandler::sequenceSizeHeader` lambda is null.
However that is not the case.
Here we are talking about result of `message.getHeaders().get(IntegrationMessageHeaderAccessor.SEQUENCE_SIZE, Integer.class)`.
Therefore it is `null` for function invocation.
Hence, has to be like this:
```
private Function<Message<?>, @Nullable Integer> windowSizeFunction = FluxAggregatorMessageHandler::sequenceSizeHeader;
```

I'm not sure why have you made this property as `public`, though...

artembilan commented on spring-integration-core/src/main/java/org/springframework/integration/aggregator/MessageListProcessor.java:
```diff
@@ -27,6 +29,7 @@
 @FunctionalInterface
 public interface MessageListProcessor {
 
+	@Nullable
```
I think this one has to be non-null as well.

artembilan reviewed (changes requested):
This is very cool!
Let's see if we can agree on my feedback.
Thanks

cppwfs commented on spring-integration-core/src/main/java/org/springframework/integration/aggregator/FluxAggregatorMessageHandler.java:
```diff
@@ -99,13 +109,15 @@ private Flux<Message<?>> releaseBy(Flux<Message<?>> groupFlux) {
 				.flatMap((windowFlux) -> windowFlux.transform(this.combineFunction));
 	}
 
+	@NullUnmarked
```
Even after the change has been applied the following line in the method is getting flagged:
https://github.com/spring-projects/spring-integration/blob/main/spring-integration-core/src/main/java/org/springframework/integration/aggregator/FluxAggregatorMessageHandler.java#L109

This is because the ` R apply(T t);` from the Function.java has not been annotated with @Nullable.

cppwfs committed:
Apply code changes from code review

artembilan commented on spring-integration-core/src/main/java/org/springframework/integration/aggregator/FluxAggregatorMessageHandler.java:
```diff
@@ -277,6 +289,7 @@ private Mono<Message<?>> messageForWindowFlux(Flux<Message<?>> messageFlux) {
 	}
 
+	@NullUnmarked
 	private static Integer sequenceSizeHeader(Message<?> message) {
```
The result of this method can be null, so:
```
private static @Nullable Integer sequenceSizeHeader(Message<?> message)
```
instead of `@NullUnmarked`

cppwfs replied:
With the change applied, It still exhibits the same behavior as enumerated here: https://github.com/spring-projects/spring-integration/pull/10097#discussion_r2141065674

artembilan committed:
Fix nullability in the `FluxAggregatorMessageHandler`

* This PR is not ready for merge.   There are failing tests, but wanted to capture where it is at.
* Also still having issue with the .apply() in FluxAggregatorMessageHandler
* Make `windowSizeFunction` as `Function<Message<?>, @Nullable Integer>`
because `sequenceSizeHeader()` may return `null` from message headers
* Extract local `subscriptionToDispose` in the `stop()` to satisfy null check context
* Use `Objects.requireNonNull(signal.get())` to satisfy `Function.apply()` contract.
The `if (signal.hasValue()) {` does the trick for us, but currently that is not visible
for that `signal.get()`
* Remove `@NullUnmarked` since we have just mitigated all the null problems

Updated tests so that they work with nullability changes

Update the tests so that they will pass with nullify changes

artembilan reviewed (changes requested):
See my last commit into your PR for the `FluxAggregatorMessageHandler` fix.
At least that what I think about those nulls.

cppwfs committed:
Resolve second set of test failures

* Remove Nullunmarked from FluxAggregatorMessageHandler
* Add TaskScheuler to tests that are failing because of nullability
* In the past the TaskScheduler attribute in the testscould be set to null and the tests would succeed.
But because of nullability additions these need to be populated

cppwfs commented on spring-integration-syslog/src/test/java/org/springframework/integration/syslog/inbound/SyslogReceivingChannelAdapterTests.java:
```diff
@@ -272,4 +263,12 @@ public void testUdpRFC5424() throws Exception {
 		adapter.stop();
 	}
 
+	private BeanFactory getBeanFactory() {
```
Some tests needed the bean factory with Mocked components.

cppwfs committed:
Add TaskScheduler for adapter in testNet tests

artembilan commented on spring-integration-core/src/main/java/org/springframework/integration/aggregator/AbstractCorrelatingMessageHandler.java:
```diff
@@ -929,6 +937,7 @@ protected Collection<Message<?>> completeGroup(Message<?> message, Object correl
 			this.logger.debug(() -> "Completing group with correlationKey [" + correlationKey + "]");
 
 			result = this.outputProcessor.processMessageGroup(group);
+			Assert.notNull(result, "the group returned a null result");
```
I think that is processor is responsible for null result, not group. But again: it really can return null. Might be something is off with our logic here?
Not saying to do that right now with such a huge pr: let’s just have a note to come back to this method!
For now, fix, please, this error message .

artembilan commented on spring-integration-core/src/main/java/org/springframework/integration/aggregator/ExpressionEvaluatingMessageListProcessor.java:
```diff
@@ -102,7 +105,9 @@ public void setExpectedType(Class<?> expectedType) {
 	public Object process(Collection<? extends Message<?>> messages) {
-		return this.evaluateExpression(this.expression, messages, this.expectedType);
+		Object object = this.evaluateExpression(this.expression, messages, this.expectedType);
+		Assert.state(object != null, "Failed to evaluate expression: " + this.expression);
```
I don’t think it is really correct . The expression has been just evaluated to null, but we don’t expect null here. Let’s say exactly that condition: the expression evaluated to null and therefore this exception .

artembilan commented on spring-integration-core/src/main/java/org/springframework/integration/aggregator/ExpressionEvaluatingMessageGroupProcessor.java:
```diff
@@ -59,7 +60,9 @@ public void setExpectedType(Class<?> expectedType) {
 	protected Object aggregatePayloads(MessageGroup group, Map<String, Object> headers) {
-		return this.processor.process(group.getMessages());
+		Object object = this.processor.process(group.getMessages());
+		Assert.notNull(object, "Result from processor must not be null");
```
Use `state()` instead. This really not an illegal arg.

artembilan commented on spring-integration-core/src/main/java/org/springframework/integration/context/IntegrationContextUtils.java:
```diff
@@ -128,6 +131,7 @@ public static MetadataStore getMetadataStore(BeanFactory beanFactory) {
 	 */
+	@Nullable
 	public static MessageChannel getErrorChannel(BeanFactory beanFactory) {
 		return getBeanOfType(beanFactory, ERROR_CHANNEL_BEAN_NAME, MessageChannel.class);
```
Was not addressed

artembilan commented on spring-integration-core/src/main/java/org/springframework/integration/context/IntegrationObjectSupport.java:
```diff
@@ -208,6 +224,7 @@ public final void setPrimaryExpression(Expression expression) {
 	}
 
 	@Override
+	@SuppressWarnings("NullAway")
```
We need to revise this latter.

artembilan reviewed (changes requested):
This is cool!
We can chat on Monday if you wish.
I also hope you do not update Copyright year manually . We have a Gradle task to do that for us 😉

cppwfs committed:
Update code based on second review.

* Added style to IDE to resolve style issues
* Resolved the missed remove from IntegrationContextUtils.
* Made sure that private test methods were marked static
* Updated Assert.NotNulls to Assert.state and updated their messages
* Rebased and updated nullification for changes

artembilan reviewed (changes requested):
Sorry, there are still some review feedback.
But this is great to be merged soon enough.

Thanks

artembilan commented:
Hey, @cppwfs !

I'm sorry to cause more headache, but we have just pushed a breaking change which might not merge clean into your branch.
That one is about changing Copyright years from let's say `2020-2025` to `2020-present`.
Let's me know and I may help you to mitigate the pain!

artembilan commented:
Consider to rebase to the latest `main`.
Have just pushed redundant `grgit` Gradle plugin usage.
Since we use now `-present` pattern for Copyright, we don't need to track modified files anymore.
That should speed up a local development loop a little bit 😄 .
